src/gateway/routers/RouterAggregator.sol

```diff
@@ -0,0 +1,231 @@
+    event File(bytes32 indexed what, address[] routers);
+
+    constructor(address gateway_) {
+        gateway = GatewayLike(gateway_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, address[] calldata routers_) external auth {
+        if (what == "routers") {
+            require(routers_.length <= MAX_ROUTER_COUNT, "RouterAggregator/exceeds-max-router-count");
+
+            // Enable new routers, setting quorum to number of routers
+            uint8 quorum_ = uint8(routers_.length);
+            for (uint8 i; i < routers_.length; ++i) {
+                // Ids are assigned sequentially starting at 1
+                validRouters[routers_[i]] = Router(i + 1, quorum_);
+            }
+
+            // Disable old routers that weren't already overridden
+            for (uint8 j = uint8(routers_.length); j < routers.length; ++j) {
+                delete validRouters[address(routers[j])];
+            }
+
+            routers = routers_;
+        } else {
+            revert("RouterAggregator/file-unrecognized-param");
+        }
+
+        emit File(what, routers_);
+    }
+
+    // --- Incoming ---
+    /// @dev Handle incoming messages, proofs, and recoveries.
+    ///      Assumes routers ensure messages cannot be confirmed more than once.
+    function handle(bytes calldata payload) public {
+        Router memory router = validRouters[msg.sender];
+        require(router.id != 0, "RouterAggregator/invalid-router");
+        _handle(payload, router);
+    }
+
+    function _handle(bytes calldata payload, Router memory router) internal {
+        if (MessagesLib.isRecoveryMessage(payload)) {
+            require(routers.length > 1, "RouterAggregator/no-recovery-with-one-router-allowed");
+            return _handleRecovery(payload);
+        }
+
+        if (router.quorum == 1 && !MessagesLib.isMessageProof(payload)) {
+            // Special case for gas efficiency
+            gateway.handle(payload);
+            emit ExecuteMessage(payload, msg.sender);
+            return;
+        }
+
+        bytes32 messageHash;
+        ConfirmationState storage state;
+        if (MessagesLib.isMessageProof(payload)) {
+            messageHash = MessagesLib.parseMessageProof(payload);
+            state = _confirmations[messageHash];
+            state.proofs[router.id - 1]++;
+
+            emit HandleProof(messageHash, msg.sender);
+        } else {
+            messageHash = keccak256(payload);
+            state = _confirmations[messageHash];
+            state.messages[router.id - 1]++;
+
+            emit HandleMessage(payload, msg.sender);
+        }
+
+        if (state.messages.countNonZeroValues() >= 1 && state.proofs.countNonZeroValues() >= router.quorum - 1) {
+            // Reduce total message confiration count by 1, by finding the first non-zero value
+            state.messages.decreaseFirstNValues(1, 1);
```

Comment:
shouldn't the values be set to 0, instead just being decreased by 1

Reply:
Yeah this is the tricky part. This is for handling multiple parallel messages. Say there are 3 duplicate messages being sent (e.g. increase invest order by 100). 1 proof router processes them all super quickly, and 1 other proof router only processes them much slower. What will happen is that the proof count of the first router will go to 3, while the proof count of the latter router will stay at 0. Each time the proof from the other router comes in, the proof count of the first will go down by 1, down from 3 to eventually 0.